### PR TITLE
Adds "Hunted" scenario w/ tracking monster 

### PR DESCRIPTION
**Pull request by ericlimer**
#### Summary

SUMMARY: Features "Adds scenario 'Hunted' which gives players a relentless nemesis monster"

#### Purpose of change

- To add a monster which can hunt the player at an extreme distance over multiple overmaps
- To emulate the film "[It Follows](https://en.wikipedia.org/wiki/It_Follows)", where a powerful but outrunnable monster relentlessly advances on the player's exact location until it is killed. 
- To enforce a 'constant movement' playstyle much like 'nomad' trait does, with a penalty that is both harsher and surmountable.

#### Describe the solution

Players who choose the scenario 'hunted' will be assigned a forced (and otherwise unselectable) trait 'hunted' and a 'kill nemesis' mission at spawn. The 'kill nemesis' mission setup spawns a nemesis monster at a nearby OMT tile as a single-monster GROUP_NEMESIS horde of a unique mon_zombie_nemesis creature with unique horde_behavior. 

When the player suffers from the hunted trait (i.e. constantly), they signal the nemesis their exact location with a unique signal_nemesis function. The nemesis horde then advances on the player's location using a modified version of the move_hordes function (move_nemesis) which uses abs_sm coordinates to move the horde towards the player even when the player and nemesis are separated by multiple overmaps. 

If the nemesis monster is killed while spawned into the reality bubble, the "kill nemesis" mission completes and the player loses the hunted trait. If the player runs away, the nemesis monster despawns back into the overmap as a horde, and continues its pursuit until the next encounter.

This trait does _not_ require wanderspawns to function.

#### Describe alternatives you've considered

-  Faking actual overmap tracking by cribbing from the nomad trait, which tracks the player's movement in a more abstract way. Abandoned because lack of ability to account for terrain/orientation could make the nemesis feel like it is just teleporting around (because it would be).
- Making "hunted" a starting trait instead of a scenario

#### Testing

- spawned into new worlds and waited for the nemesis to arrive
- tested hordes tracking ability at a distance of 10+ overmaps
- encountered and escaped the nemesis to test its ability to de/respawn.
- spawned into new worlds, completed and failed the mission in various ways, started new characters to ensure failed missions to not leave artifacts in subsequent savegames

#### Additional context /future goals

I am a novice coder designing this feature primarily for my own use. My implementation is functional, but probably not optimal. Input is welcome and encouraged.

Some future features I'm considering: 

- multiple nemesis types (zombie, alien, nether, robotic) and an option to randomize or choose between them
- an option to have the nemesis be a random enemy of a certain difficulty tier
- an option for infinitely respawning nemesis
- balance adjustments to the current nemesis

### CODE REVIEW

File: data/json/npcs/missiondef.json [hunk 1]
```diff
@@ -49,7 +49,7 @@
     },
     "start": {
       "effect": [
-        { "npc_add_trait": "ASTHMA" },
```
Comment (Fris0uman) [inline data/json/npcs/missiondef.json:52]: Why this change?

File: data/json/npcs/missiondef.json [hunk 2]
```diff
@@ -461,6 +461,31 @@
...
+    "monster_kill_goal": 1,
+    "difficulty": 5,
+    "value": 200000,
+    "urgent": true,
+    "start": "kill_nemesis",
+    "end": { "effect": [ { "u_lose_trait": "HAS_NEMESIS" } ] },
+    "origins": [ "ORIGIN_GAME_START" ],
+    "dialogue": {
+      "describe": "The eater of the dead… something was ripping zombies to shreds and only leaving a few scattered limbs…",
+      "offer": "A few days ago another survivor and I were trying to avoid the cities by staying in the woods during the day and foraging for gear at night.  It worked well against the normal zed's but one night something caught onto our trail and chased us for ten minutes or so until we decided to split up and meet-up back here.  My buddy never showed up and I don't have the means to kill whatever it was.  Can you lend a hand?",
```
Comment (Fris0uman) [inline data/json/npcs/missiondef.json:479]: Did you copy-paste the jabberwock lines?

File: data/json/npcs/missiondef.json (hunk 1 above)
Comment (ericlimer) [reply data/json/npcs/missiondef.json:52]: Accident! Cut instead of copied. will fix.

File: data/json/npcs/missiondef.json (hunk 2 above)
Comment (ericlimer) [reply data/json/npcs/missiondef.json:479]: I did. Documentation left me under the impression these descriptions are requires, though they will never appear in game because this mission is not assigned by an NPC. 

I suppose I should modify them to "n/a" or something similar

Comment (Fris0uman) [reply data/json/npcs/missiondef.json:479]: I don't think you need them, the missions you get from starting scenarios don't have them.

Comment (ericlimer) [reply data/json/npcs/missiondef.json:479]: Looks like you are correct and I misread the documentation! Removing all the dialog options.

**Comment by Qrox** [comment]
If the player character dies and a new character with the hunted trait is created, does the old nemesis monster still track the new character?

**Comment by Gourmand-is-OP** [comment]
It needs much more health, armor, and damage. Considering it's a single monster, it should be at least as difficult as a shoggoth.

File: data/json/monsters/zed_misc.json [hunk 3]
```diff
@@ -299,6 +299,55 @@
...
+    "name": { "str": "doppelganger" },
+    "description": "A repulsive humanoid, its grotesque muscles throb with an infernal energy. In its twisted visage, which has haunted your dreams since childhood, you can see the demonic reflection of your own face.",
+    "default_faction": "zombie",
+    "bodytype": "human",
+    "species": [ "ZOMBIE", "HUMAN", "NEMESIS" ],
+    "diff": 5,
+    "volume": "875000 ml",
+    "weight": "200 kg",
+    "hp": 480,
+    "speed": 50,
```
Comment (anothersimulacrum) [inline data/json/monsters/zed_misc.json:314]: This seems quite slow. Maybe bump it up to 70 or 80? Still outrunnable, but not quite so trivial.

File: data/json/monsters/zed_misc.json [hunk 4]
```diff
@@ -299,6 +299,55 @@
...
+    "armor_bash": 8,
+    "armor_cut": 12,
+    "armor_bullet": 10,
+    "vision_day": 50,
+    "vision_night": 4,
+    "harvest": "zombie",
+    "special_attacks": [ [ "SMASH", 20 ] ],
+    "death_drops": "mon_zombie_hulk_death_drops",
+    "death_function": [ "NORMAL" ],
+    "flags": [
```
Comment (anothersimulacrum) [inline data/json/monsters/zed_misc.json:333]: Consider some sort of regen.

File: src/game.cpp [hunk 5]
```diff
@@ -813,6 +814,13 @@ bool game::start_game()
         new_mission->assign( u );
     }
 
+    // Assign nemesis kill mission if character has the 'hunted' trait
+    if( u.has_trait( trait_id( "HAS_NEMESIS" ) ) ) {
```
Comment (anothersimulacrum) [inline src/game.cpp:818]: Why are you adding the static version if you're not using it here?

File: src/mission_start.cpp [hunk 6]
```diff
@@ -125,6 +125,14 @@ void mission_start::kill_horde_master( mission *miss )
     tile.save();
 }
 
+void mission_start::kill_nemesis( mission *miss )
+{
+    // Pick an area for the nemesis to spawn
+    const auto center = get_player_character().global_omt_location();
```
Comment (anothersimulacrum) [inline src/mission_start.cpp:131]: Could you replace auto here with the proper type?

File: src/overmap.cpp [hunk 7]
```diff
@@ -2137,6 +2142,20 @@ void overmap::signal_hordes( const tripoint_rel_sm &p_rel, const int sig_power )
     }
 }
 
+void overmap::signal_nemesis( const tripoint_rel_sm &p_rel, const int sig_power )
+{
+    tripoint_om_sm p( p_rel.raw() );
+    for( auto &elem : zg ) {
```
Comment (anothersimulacrum) [inline src/overmap.cpp:2148]: Same here, please replace the auto.

File: src/overmapbuffer.cpp [hunk 8]
```diff
@@ -486,6 +486,27 @@ void overmapbuffer::signal_hordes( const tripoint_abs_sm &center, const int sig_
     }
 }
 
+void overmapbuffer::signal_nemesis( const tripoint_abs_sm &p, const int sig_power )
+{
+
+    const auto radius = sig_power;
```
Comment (anothersimulacrum) [inline src/overmapbuffer.cpp:492]: int, not auto

File: src/overmapbuffer.cpp [hunk 9]
```diff
@@ -486,6 +486,27 @@ void overmapbuffer::signal_hordes( const tripoint_abs_sm &center, const int sig_
...
+    for( auto &om : get_overmaps_near( p, radius ) ) {
```
Comment (anothersimulacrum) [inline src/overmapbuffer.cpp:493]: Presumably this auto here should be overmap?

**Comment by Maleclypse** [comment]
I would happily use this with some reflavoring in Aftershock and I think Magiclysm could have some uses for this as well.

**Comment by John-Candlebury** [comment]
The hounds of tindalos can use this as well. Although it would make the flaming eyes even more dangerous.

**Comment by Night-Pryanik** [comment]
I suppose this trait would be better limited to a new separate scenario. Like Medieval Peasant scenario limits the use of "Churl" trait. A separate scenario could perfectly justify some supernatural stuff.

**Comment by int-ua** [comment]
Sounds interesting. How does it behave if you reach it while underground? Does it spawn on the surface?

**Comment by eilaattwood** [comment]
Looks like a job for the ultimate Cataclysm weapon - car.

**Comment by eilaattwood** [comment]
> I'm not sure this kind of supernatural flavour fits within the vanilla lore.

Just replace supernatural monster with some kind of police robot, and surepnatural flavour is gone.
Btw, this scenario gives me strong Two-Handed Engine (by Henry Kuttner) vibes.

**Comment by DrPariah** [comment]
I really, REALLY like this idea. And it's no more 'supernatural' than anything in the base game is. The Blob, and Nether entities in general, are completely alien to human thought processes. You don't ask why the monster has decided that you, specifically, need to go - you ask _why not_?
Although, thinking about it, it makes sense to have a selection of potential stalkers, rather than locking it in to just one. That way, it can fit in with people's particular RP. You could have a selection menu appear like you do when you select a Martial Arts trait, and choose the form of your destructor. Offering options like Robot, Nether, "Zombie" (actually a high-functioning Feral Human, but whatever, not much difference), Mutated Abomination (you don't know who or what this used to be, but it seems to hold a grudge), possibly others. Then you could tie into the architecture set up here to allow other mods like Magiclysm and Aftershock to hold their own, respective versions in the same way Mythical Martial Arts hooks into the mainline martial arts system.

**Comment by ericlimer** [comment]
I'm picking this project back up!

**Comment by Zireael07** [comment]
It's so old that it wouldn't be anywhere close to mergeable even if it were reopened.

**Comment by ericlimer** [comment]
I was able to merge the year's worth of upstream changes to this branch almost entirely without breaking the work I'd done, so I think I should be able to get it fully up to speed once I tie up a few loose threads and commit all those updates. 

If that's gonna make a big mess of this PR though I can just open another one instead. Just thought it'd be fun to preserve this for posterity if possible.

File: src/overmapbuffer.cpp (hunk 9 above)
Comment (ericlimer) [reply src/overmapbuffer.cpp:495]: i'm actually struggling with figuring out what this auto should be; i will admit here to just having copied the loop code from the existing "signal_hordes" function, knowing that it serves my purposes but not precisely how it works.

Comment (ericlimer) [reply src/overmapbuffer.cpp:495]: nevermind! i figured this out 😄

**Comment by ericlimer** [comment]
Finally got this bad boy up and running! This is the biggest project I've attempted so far so apologies in advance for any stupid newbie mistakes or issues. Happy to address them!

File: data/json/mutations/mutations.json [hunk 10]
```diff
@@ -7058,5 +7058,13 @@
     "description": "You have fast reflexes, allowing you to dodge attacks and grabs more easily.",
     "starting_trait": true,
     "dodge_modifier": 2
+  },
+  {
+    "type": "mutation",
+    "id": "HAS_NEMESIS",
+    "name": { "str": "Hunted" },
+    "points": -2,
+    "description": "Since the start of the Cataclysm, something has been hunting you relentlessly. You can outrun it, but it always seems to find you eventually. The only escape is for one of you to die.",
```
Comment (actual-nh) [inline data/json/mutations/mutations.json:7067]:
```suggestion
    "description": "Since the start of the Cataclysm, something has been hunting you relentlessly.  You can outrun it, but it always seems to find you eventually.  The only escape is for one of you to die.",
```

File: data/json/npcs/missiondef.json [hunk 11]
```diff
@@ -475,6 +475,18 @@
       "failure": "Quitting already?"
     }
   },
+  {
+    "id": "MISSION_KILL_NEMESIS",
+    "type": "mission_definition",
+    "name": { "str": "Kill nemesis" },
+    "description": "Since the start of the Cataclysm, something has been hunting you relentlessly. You can outrun it but never truly escape it. One of you must die.",
```
Comment (actual-nh) [inline data/json/npcs/missiondef.json:482]:
```suggestion
    "description": "Since the start of the Cataclysm, something has been hunting you relentlessly.  You can outrun it but never truly escape it.  One of you must die.",
```

**Review by actual-nh — COMMENTED** [review]
The dialogue style for CDDA requires two spaces after each period.

File: data/json/species.json [hunk 12]
```diff
@@ -133,6 +133,12 @@
     "footsteps": "shuffling.",
     "bleeds": "fd_blood"
   },
+  {
+    "type": "SPECIES",
+    "id": "NEMESIS",
+    "description": "a zombie",
+    "footsteps": "shuffling."
+  },
```
Comment (Night-Pryanik) [inline data/json/species.json:141]: Should it bleed?

File: src/suffer.cpp [hunk 13]
```diff
@@ -326,6 +327,11 @@ void Character::suffer_while_awake( const int current_stim )
         vomit();
     }
 
+    if( has_trait( trait_HAS_NEMESIS ) && one_turn_in( 2_minutes ) ) {
+        signal_nemesis();
+    }
+
+
```
Comment (Night-Pryanik) [inline src/suffer.cpp:334]:
```suggestion
```

File: data/json/species.json (hunk 12 above)
Comment (ericlimer) [reply data/json/species.json:141]: moot point it turns out! i'm realizing now that the 'nemesis' species isnt actually necessary to the implementation i wound up with, so i think im going to remove this entirely.

**Comment by ericlimer** [comment]
Gonna take a minute here to address some valid questions and concerns folks raised much earlier in this PR's development:

@Fris0uman said:
> I'm not sure this kind of supernatural flavour fits within the vanilla lore. Also what does it do when the player does not have the trait?

Fair enough! I've restyled the monster to a hulk-inspired zombie called "relentless hulk" to remove demonic/supernatural elements and better fit the flavor of the vanilla game. 

The monster only spawns once at world-gen, and only if the player has chosen the 'hunted' trait in character creation which triggers the Kill_Nemesis mission that spawns it. It does not spawn naturally and players without the trait should never encounter it. 

@Qrox asked:
> If the player character dies and a new character with the hunted trait is created, does the old nemesis monster still track the new character?

No. If a player with the hunted trait dies without having first killed the nemesis, the hunted trait triggers a "remove_nemesis" method in the end-game cleanup that removes the nemesis monster. A new character with the hunted trait will spawn a new nemesis monster to track them.

@Gourmand-is-OP commented: 
> It needs much more health, armor, and damage. Considering it's a single monster, it should be at least as difficult as a shoggoth.

I've adjusted the balance a little since then, but am still happy to hear thoughts. My goal is to have it defeatable with difficulty by a late early-game or early mid-game player. 

@eilaattwood said:
> Looks like a job for the ultimate Cataclysm weapon - car.

Yeah this is by far my biggest balance concern, but I'm not sure how to mitigate it. I found [this potential solution from a redditor ](https://www.reddit.com/r/Cataclysm_DDA/comments/ozfmxf/making_a_single_monster_resilient_against_the/h8904bb/?utm_source=reddit&utm_medium=web2x&context=3)intriguing but i haven't looked into implementing it.

@int-ua asked: 
> Sounds interesting. How does it behave if you reach it while underground? Does it spawn on the surface? At least at first I would've liked to be able to tune its' speed. Also, an in-game event that triggers this monster sounds interesting to me.

For now, it only spawns on the surface z level. If the player is above or below ground, it will lurk on the surface of the same overmap. This is in part bc I'm not sure how to make it spawn in elsewhere, but also limits the impression that it teleports into impossible spaces. 

@DrPariah suggested: 

> You could have a selection menu appear like you do when you select a Martial Arts trait, and choose the form of your destructor. Offering options like Robot, Nether, "Zombie" (actually a high-functioning Feral Human, but whatever, not much difference), Mutated Abomination (you don't know who or what this used to be, but it seems to hold a grudge), possibly others. Then you could tie into the architecture set up here to allow other mods like Magiclysm and Aftershock to hold their own, respective versions in the same way Mythical Martial Arts hooks into the mainline martial arts system.
EDIT: Optionally, would it be possible to have a Relentless Hunter? One that respawns after a given length of time after you've killed it? Though that does have the potential to become annoying after a while.

I like all these ideas and will consider implementing them if this PR is accepted. Mod hooks are a little beyond the scope my interest or skill, but otherwise these options seem feasible.

**Comment by anothersimulacrum** [comment]
I agree with Night-Pryanik - this fits perfectly well in a scenario.

**Comment by ericlimer** [comment]
Sure. I see the way traits, hobbies and professions have been separated makes this an odd fit as a trait. I initially chose that implementation chiefly so it could be used in combination with different scenarios, but I'm not married to it. 

I'll look into changing it to a scenario and maybe bringing back some of the original flavoring if that's less of an issue in a scenario form.

**Comment by ericlimer** [comment]
I've changed this feature from a trait to a scenario, as recommended by @Night-Pryanik and @anothersimulacrum.

**Review by Maleclypse — APPROVED** [review]
The C++ is above me but the JSON looks good and even better news that with "copy-from" mods could edit what the nemesis is.

File: data/json/monsters/zed_misc.json [hunk 14]
```diff
@@ -442,6 +442,57 @@
...
+  {
+    "id": "mon_zombie_nemesis",
+    "type": "MONSTER",
+    "name": { "str": "relentless hulk" },
+    "description": "A hulking zombie swollen to the size of six men.  Its inky black eyes are locked on you.  No matter how far you run, it always seems to find you.",
+    "default_faction": "zombie",
+    "bodytype": "human",
+    "species": [ "ZOMBIE", "HUMAN" ],
+    "diff": 5,
+    "volume": "875000 ml",
```
Comment (CountAlex) [inline data/json/monsters/zed_misc.json:454]: It seems to me that you need to balance volume/weight balance as for now the density of a monster is <0.25, and that's just weird. Not enough to float, but that's barely different from aerogel.

Comment (ericlimer) [reply data/json/monsters/zed_misc.json:454]: Good catch, I can make some adjustments. Worth noting these values are copied from the skeletal juggernaut, where they would make a little more sense if it was a doot doot-style cartoon skeleton, but considering it's stylized as bones and flesh, that might warrant a change as well.
